src/Hamiltonians/GutzwillerSampling.jl

```diff
@@ -110,3 +120,57 @@ function Base.getindex(h::GutzwillerOffdiagonals{F,T,A,G}, i)::Tuple{F,T} where
 end
 
 Base.size(h::GutzwillerOffdiagonals) = size(h.offdiagonals)
+
+"""
+    TransformUndoer(k::GutzwillerSampling, op::AbstractHamiltonian)
+    TransformUndoer(k::GutzwillerSampling)
+
+For a Gutzwiller similarity transformation ``\\hat{G} = f \\hat{H} f^{-1}`` define the operator 
+``f^{-1} A f^{-1}``, and special case ``f^{-2}``, in order to calculate observables. 
+    
+See [`AllOverlaps`](@ref), [`GutzwillerSampling`](@ref).
+"""
+function TransformUndoer(k::GutzwillerSampling, op::AbstractHamiltonian)
+    T = promote_type(eltype(k), eltype(op))
+    return TransformUndoer{T,typeof(k),typeof(op)}(k, op)
+end
+
+LOStructure(::Type{<:TransformUndoer{<:Any,<:GutzwillerSampling,A}}) where {A} = LOStructure(A)
+
+function LinearAlgebra.adjoint(s::TransformUndoer{T,<:GutzwillerSampling,<:AbstractHamiltonian}) where {T}
+    a_adj = adjoint(s.op)
+    return TransformUndoer{T,typeof(s.transform),typeof(a_adj)}(s.transform, a_adj)
+end
+
+function diagonal_element(s::TransformUndoer{<:Any,<:GutzwillerSampling,<:AbstractHamiltonian}, add)
+    diagH = diagonal_element(s.transform.hamiltonian, add)
+    diagA = diagonal_element(s.op, add)
+    return gutzwiller_modify(diagA, true, s.transform.g, 0., 2 * diagH)
+end
+
+function num_offdiagonals(s::TransformUndoer{<:Any,<:GutzwillerSampling,<:Any}, add)
+    return num_offdiagonals(s.op, add)
+end
+
+function get_offdiagonal(s::TransformUndoer{<:Any,<:GutzwillerSampling,<:Any}, add, chosen)
+    newadd, offd = get_offdiagonal(s.op, add, chosen)
+    # Gutzwiller `f` operator is diagonal
+    diagH1 = diagonal_element(s.transform.hamiltonian, add)
+    diagH2 = diagonal_element(s.transform.hamiltonian, newadd)
+    return newadd, gutzwiller_modify(offd, true, s.transform.g, 0., diagH1 + diagH2)
+end
+
+# Special case `f^2`
+function TransformUndoer(k::GutzwillerSampling)
+    T = eltype(k)
+    return TransformUndoer{T,typeof(k),Nothing}(k,nothing)
+end
```

Comment:
the generic method is designed to throw an error if an unsupported transformation is attempted

Reply:
I also realised that it could be nice to check if the Hamiltonian transformation is consistent with the (un)transformed operators. For example, if `G = GutzwillerSampling(H,g)` is passed to `lomc!` then the same `G` is passed to `AllOverlaps`. I added a function to check this called `check_transform`. It is a little bit clunky but could be a good addition.